chain/consensus/mir/mine.go

```diff
@@ -0,0 +1,185 @@
+package mir
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/v1api"
+	"github.com/filecoin-project/lotus/chain/consensus/common"
+	"github.com/filecoin-project/lotus/chain/types"
+	ltypes "github.com/filecoin-project/lotus/chain/types"
+	mirTypes "github.com/filecoin-project/mir/pkg/types"
+)
+
+// Mine mines Filecoin blocks.
+
+// Mine implements the following algorithm:
+// 1. Retrieve messages and cross-messages from mempool.
+// 2. Send these messages to the Mir node.
+// 3. Receive ordered messages from the Mir node and push them into the next Filecoin block.
+// 4. Submit this block.
+func Mine(ctx context.Context, miner address.Address, api v1api.FullNode) error {
+	log.Info("Mir miner started")
+	defer log.Info("Mir miner stopped")
+
+	// TODO: Suppose we want to use Mir in Root and in a subnet at the same time.
+	// Do we need two Mir agents for that?
+	// nodeID, err := api.ID(ctx)
+	// if err != nil {
+	//	log.Fatalf("unable to get a node ID: %s", err)
+	// }
+	nodeID := NodeID()
+	nodeNumber := NodeNumber()
+
+	mirAgent, err := NewMirAgent(nodeID, nodeNumber)
+	if err != nil {
+		return err
+	}
+	mirErrors := mirAgent.Start(ctx)
+	mirHead := mirAgent.App.ChainNotify
+
+	netName, nerr := api.StateNetworkName(ctx)
+	if nerr != nil {
+		return nerr
+	}
+	subnetID := address.SubnetID(netName)
+
+	log.Infof("Mir miner params: network name - %s, subnet ID - %s", netName, subnetID)
+
+	submit := time.NewTicker(SubmitInterval)
+	defer submit.Stop()
+
+	for {
+		base, err := api.ChainHead(ctx)
+		if err != nil {
+			log.Errorw("failed to get the head of chain", "error", err)
+			continue
+		}
+
+		select {
+		case <-ctx.Done():
+			log.Debug("Mir miner: context closed")
+			return nil
+		case merr := <-mirErrors:
```

Comment:
I like this approach to kill the mining process. Do you think it make sense to propagate it to all consensus algorithms?

Context: As we discussed sync, our consensus algorithms are architected a bit differently to how `filcns` is, and for subnets we don't run miners in a separate process. This would allow us to terminate the mining process if there is an error instead of looping with errors forever (my only concern with this would be if there are recoverable errors that don't lead to an infinite looping error in mining). I leave the final decision to you :)

Reply:
Created an issue https://github.com/filecoin-project/eudico/issues/188